### PR TITLE
Nd Array Support and Dot Product Operation added

### PR DESCRIPTION
**Pull request by karanchahal**
Now Tensor's support arrays of various shapes to be initialised. Also dot product multiplication between tensors is supported. The backward pass is left to be done and so are the tests.
Lots of changes to the overall structure of ops and tensors. Hence this is a big change. I'll put the Backward op and test in other pull requests.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
 .vscode
 main
-build
\ No newline at end of file
+build
+a.out
+ds/
```
Comment (uditarora) [inline .gitignore]: nit: newline at the end of file

File: dotOperation.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#include "operation.h"
+
+#ifndef __OP_DOT_INCLUDED__  
+#define __OP_DOT_INCLUDED__  
+template <typename T>
+class DotOperation : public Operation<T> {
+
```
Comment (uditarora) [inline dotOperation.h]: nit: remove this newline for consistency

File: dotOperation.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    public:
+        Tensor<T> *t1=NULL, *t2=NULL; // Input tensors
+        Tensor<T> *t3=NULL; // Output tensor
+    
+    DotOperation(Tensor<T> *t1, Tensor<T> *t2) {
+        this->t1 = t1;
+        this->t2 = t2;
+    }
+    void backward(Matrix<T> grad);
```
Comment (uditarora) [inline dotOperation.h:16]: nit: add newline before each new function definition

File: dotOperation.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    Tensor<T> forward();
+};
+
+#endif
```
Comment (uditarora) [inline dotOperation.h]: nit: newline at end of file

File: matrix.h [hunk 5]
```diff
@@ -63,7 +63,7 @@ struct Matrix{
             // Both of these matrices should have at least 2 dimensions
             
             // Condition for n and m both 3D matrices or greater than 3D matrices or 3d and 1d matrix
```
Comment (uditarora) [inline matrix.h]: The comment here doesn't seem to match the condition being checked

File: addOperation.h [hunk 6]
```diff
@@ -11,15 +11,17 @@ two tensors.
...
+        Tensor<T> *t3=NULL; // Output tensor
     
-    AddOperation(Tensor<T> *t1, Tensor<T> *t2) {
+    AddOperation(Tensor<T> *t1, Tensor<T> *t2): Operation<T>(t1,t2){
         this->t1 = t1;
         this->t2 = t2;
     }
+    void backward(Matrix<T> grad);
 
     void backward(vector<T> grad);
```
Comment (uditarora) [inline addOperation.h]: remove, redundant

File: addOperation.h [hunk 7]
```diff
@@ -11,15 +11,17 @@ two tensors.
...
-        Tensor<T> *t3; // Output tensor
+        Tensor<T> *t1=NULL, *t2=NULL; // Input tensors
+        Tensor<T> *t3=NULL; // Output tensor
     
-    AddOperation(Tensor<T> *t1, Tensor<T> *t2) {
+    AddOperation(Tensor<T> *t1, Tensor<T> *t2): Operation<T>(t1,t2){
         this->t1 = t1;
         this->t2 = t2;
     }
+    void backward(Matrix<T> grad);
```
Comment (uditarora) [inline addOperation.h:22]: nit: newline above

File: common.h [hunk 8]
```diff
@@ -0,0 +1,5 @@
+#include<iostream>
+#include<vector>
+#include<assert.h>
+#include<memory>
+using namespace std;
```
Comment (uditarora) [inline common.h]: nit: newline at end, and space before angular braces and newline after all include statements

File: divideOperation.h [hunk 9]
```diff
@@ -11,13 +11,14 @@ two tensors.
 template <typename T>
 class DivideOperation : public Operation<T> {
     public:
-        Tensor<T> *t1, *t2; // Input tensors
-        Tensor<T> *t3; // Output tensor
+        Tensor<T> *t1=NULL, *t2=NULL; // Input tensors
```
Comment (uditarora) [inline divideOperation.h]: nit: whitespace before and after '=' sign

File: addOperation.h [hunk 10]
```diff
@@ -11,15 +11,17 @@ two tensors.
 template <typename T>
 class AddOperation : public Operation<T> {
     public:
-        Tensor<T> *t1, *t2; // Input tensors
-        Tensor<T> *t3; // Output tensor
+        Tensor<T> *t1=NULL, *t2=NULL; // Input tensors
+        Tensor<T> *t3=NULL; // Output tensor
     
-    AddOperation(Tensor<T> *t1, Tensor<T> *t2) {
+    AddOperation(Tensor<T> *t1, Tensor<T> *t2): Operation<T>(t1,t2){
```
Comment (uditarora) [inline addOperation.h]: nit: whitespace before ':' everywhere

File: dotOperation.h [hunk 11]
```diff
@@ -0,0 +1,21 @@
+#include "operation.h"
+
+#ifndef __OP_DOT_INCLUDED__  
+#define __OP_DOT_INCLUDED__  
+template <typename T>
```
Comment (uditarora) [inline dotOperation.h:6]: nit: newline above

File: dotOperation.h [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+class DotOperation : public Operation<T> {
+
+    public:
+        Tensor<T> *t1=NULL, *t2=NULL; // Input tensors
```
Comment (uditarora) [inline dotOperation.h]: nit: unindent these member variables across all classes

File: matrix.h [hunk 13]
```diff
@@ -0,0 +1,268 @@
+#include "common.h"
+#include "overloads.h"
+
+#ifndef __MATRIX_FLOAT_INCLUDED__   
+#define __MATRIX_FLOAT_INCLUDED__  
+
+
```
Comment (uditarora) [inline matrix.h]: nit: remove extra newline

File: matrix.h [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+                }
+            }
+        }
+
+        return true;
+    }
+
+    // Assumes result vector is filled with 0's in the beginning
+    // Performs 2D matrix multiplication
+    void dot(vector<T> &res,
```
Comment (uditarora) [inline matrix.h]: nit: describe params

File: matrix.h [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+                }
+                res[(startRes+ i*col2 + k)] = sum;
+            }
+        }
+    }
+
+    // Matrix Multiply of any big matrix with a 2D matrix
+    // Only the inner product will be computed
+    // Assumes matrices are of right size, so no validations
+    void mmultiply(vector<T> &res,
```
Comment (uditarora) [inline matrix.h]: nit: change to matmul like numpy

File: matrix.h [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+        }
+    }
+
+    public:
+    vector<T> val;
+    vector<int> shape; // shape will be a vector 
+    
+    Matrix() {
+        //Default
+    }
```
Comment (uditarora) [inline matrix.h:193]: nit:newline below

File: matrix.h [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+    // different matrices
+    Matrix(vector<T> val, vector<int> shape) {
+        // Deep Copy
+        assert("Shape and size of vector are incompatible !" && verifyShape(val,shape));
+        this->val = val;
+        this->shape = shape;
+        this->elems = computeShapes(shape);
+    }
+
+
```
Comment (uditarora) [inline matrix.h]: nit: remove extra newline

File: matrix.h [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+        }
+        out<<" ]";
+        return out;
+    }
+
+    template<typename E>
+    friend ostream & operator << (ostream &out, Matrix<E> &m);
+
+    template<typename E>
+    friend Matrix<E> operator + (const Matrix<E> &lhs, const Matrix<E> &rhs);
```
Comment (uditarora) [inline matrix.h]: Redundant, remove

File: matrix.h [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+        return Matrix(res, resShape);
+    }
+
+    Matrix<T> exp() {
+        auto res = exponent(this->val);
+        auto resShape = this->shape;
+        return Matrix(res, resShape);
+    }
+
+    // For now lhs can be really big but rhs will be 2D matrix.
```
Comment (uditarora) [inline matrix.h]: nit: add comment explaining what this operator does

File: matrix.h [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+};
+
+template<typename T>
+ostream & operator << (ostream &out, Matrix<T> &m) {
+    vector<int> stack;
+    return m.print(out,stack,0);
+}
+
+template<typename E>
+Matrix<E> operator / (const E e,const Matrix<E> &rhs) {
```
Comment (uditarora) [inline matrix.h]: nit: whitespace after ','

**Comment by karanchahal** [comment]
Hey @uditarora , I updated the code with your suggestions and added comments. Hope this is fine :)

**Review by uditarora — APPROVED** [review]

**Pull request merged**